r/src/array__to_vector.cpp

```diff
@@ -0,0 +1,827 @@
+}
+
+template <int RTYPE>
+struct Converter_SimpleArray {
+  using Vector = Rcpp::Vector<RTYPE, Rcpp::NoProtectStorage>;
+
+  static SEXP Allocate(R_xlen_t n, const ArrayVector&) { return Vector(no_init(n)); }
+
+  static Status Ingest(SEXP data_, const std::shared_ptr<arrow::Array>& array,
+                       R_xlen_t start, R_xlen_t n) {
+    Vector data(data_);
+    using value_type = typename Vector::stored_type;
+    auto null_count = array->null_count();
+
+    if (n == null_count) {
+      std::fill_n(data.begin() + start, n, default_value<RTYPE>());
+    } else {
+      auto p_values = array->data()->GetValues<value_type>(1);
+      if (!p_values) {
+        return Status::Invalid("Invalid data buffer");
+      }
+
+      // first copy all the data
+      std::copy_n(p_values, n, data.begin() + start);
+
+      if (null_count) {
+        // then set the sentinel NA
+        arrow::internal::BitmapReader bitmap_reader(array->null_bitmap()->data(),
+                                                    array->offset(), n);
+
+        for (size_t i = 0; i < n; i++, bitmap_reader.Next()) {
+          if (bitmap_reader.IsNotSet()) {
+            data[i + start] = default_value<RTYPE>();
+          }
+        }
+      }
+    }
+    return Status::OK();
+  }
+};
+
+struct Converter_Date32 {
+  static SEXP Allocate(R_xlen_t n, const ArrayVector&) {
+    IntegerVector data(no_init(n));
+    data.attr("class") = "Date";
+    return data;
+  }
+
+  static Status Ingest(SEXP data_, const std::shared_ptr<arrow::Array>& array,
+                       R_xlen_t start, R_xlen_t n) {
+    return Converter_SimpleArray<INTSXP>::Ingest(data_, array, start, n);
+  }
+};
+
+struct Converter_String {
+  static SEXP Allocate(R_xlen_t n, const ArrayVector&) {
+    return StringVector_(no_init(n));
+  }
+
+  static Status Ingest(SEXP data_, const std::shared_ptr<arrow::Array>& array,
+                       R_xlen_t start, R_xlen_t n) {
+    StringVector_ data(data_);
+    auto null_count = array->null_count();
+
+    if (null_count == n) {
+      std::fill_n(data.begin(), n, NA_STRING);
+    } else {
+      auto p_offset = array->data()->GetValues<int32_t>(1);
+      if (!p_offset) {
+        return Status::Invalid("Invalid offset buffer");
+      }
+      auto p_data = array->data()->GetValues<char>(2, *p_offset);
+      if (!p_data) {
+        // There is an offset buffer, but the data buffer is null
+        // There is at least one value in the array and not all the values are null
+        // That means all values are empty strings so there is nothing to do
+        return Status::OK();
+      }
+
+      if (null_count) {
+        // need to watch for nulls
+        arrow::internal::BitmapReader null_reader(array->null_bitmap_data(),
+                                                  array->offset(), n);
+        for (int i = 0; i < n; i++, null_reader.Next()) {
+          if (null_reader.IsSet()) {
+            auto diff = p_offset[i + 1] - p_offset[i];
```

Comment:
I think it would be worth using the `StringArray::GetView` since it returns you a `util::string_view` which lowers the maintenance on bookeeping `p_data`.